File: remove_rom.pl
#!/usr/bin/perl -w

# First 4K of Pac-Man ROM
$file = 'pacman.6e';
open FILE, "<$file" and binmode FILE or die "$file: $!\n";
read FILE, $rom='', -s $file;
close FILE;

# Full TAP image for emulator, including ROM
$file = 'pacemuzx.tap';
open FILE, "<$file" and binmode FILE or die "$file: $!\n";
read FILE, $data='', -s $file;
close FILE;

$index = index($data, $rom);
die "ROM image not found in TAP file!\n" if $index < 0;

$file = 'start.part';
open FILE, ">$file" and binmode FILE or die "$file: $!\n";
print FILE substr $data, 0, $index;
close FILE;

$file = 'end.part';
open FILE, ">$file" and binmode FILE or die "$file: $!\n";
print FILE substr $data, $index+16384;
close FILE;

File: remove_rom.py
#!/usr/bin/env python
import sys
import argparse

parser = argparse.ArgumentParser("Remove ROM placeholder from pacemu master disk image")
parser.add_argument("tapfile", help="input TAP file containing ROM")
parser.add_argument("romfile", help="start of ROM image to remove")
parser.add_argument("romsize", type=int, help="size of ROM to remove")
parser.add_argument("startfile", help="output file for start segment")
parser.add_argument("endfile", help="output file for end segment")
args = parser.parse_args()

with open(args.tapfile, 'rb') as tapfile:
    tap = tapfile.read()

with open(args.romfile, 'rb') as romfile:
    rom = romfile.read()

rom_offset = tap.find(rom)
if rom_offset == -1:
    sys.exit("error: ROM image not found in TAP file!")

start_part = tap[:rom_offset]
end_part = tap[rom_offset + args.romsize:]

with open(args.startfile, 'wb') as startfile:
    startfile.write(start_part)

with open(args.endfile, 'wb') as endfile:
    endfile.write(end_part)
